### PR TITLE
[ALTO] Use separate messages for TPC authentication [ALTO-9]

### PR DESCRIPTION
**Pull request by mdumandag**
We have decided to not add TPC related parameters to the stable
authentication codecs, because we are not sure about the
final TPC implementation and its requirements. Essentially,
these parameters can change in the next releases, but we
can't change them in the stable codecs because of the compatibility
guarantees we have.

Therefore, we decided to introduce the duplicate experimental client
authentication messages that would contain the TPC-related parameters.
Since the feature is BETA and we won't be providing more than the
same minor version compatibility for it, we can play with these codecs
as we want and remove them while promoting the TPC to stable
and move any additional parameters we have added to the stable codecs.

protocol PR: https://github.com/hazelcast/hazelcast-client-protocol/pull/460

### CODE REVIEW

**Comment by pveentjer** [comment]
The authentication message would be a lot more flexible if a set of key values would be returned. Currently, the position of the argument in the stream determines its meaning and this makes it inflexible.

File: hazelcast/src/main/java/com/hazelcast/client/impl/clientside/HazelcastClientInstanceImpl.java [hunk 1]
```diff
@@ -442,14 +442,13 @@ private void startHeartbeat() {
         long heartbeatInterval = properties.getPositiveMillisOrDefault(HEARTBEAT_INTERVAL);
         ILogger logger = loggingService.getLogger(HeartbeatManager.class);
         HeartbeatManager.start(this, executionService, logger,
-                heartbeatInterval, heartbeatTimeout,
-                Collections.unmodifiableCollection(connectionManager.getActiveConnections()));
+                heartbeatInterval, heartbeatTimeout, connectionManager.getActiveConnections());
```
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/clientside/HazelcastClientInstanceImpl.java:445]: any reasoning of this change?

File: hazelcast/src/main/java/com/hazelcast/client/impl/connection/ClientConnection.java [hunk 2]
```diff
@@ -52,4 +54,11 @@ public interface ClientConnection extends Connection {
     // used in tests
     Map<Long, EventHandler> getEventHandlers();
 
+    /**
+     * Returns the Alto channels associated with this connection,
+     * or {@code null}, if the client is not Alto-aware or the
+     * Alto is disabled on the server-side.
+     */
+    @Nullable
+    Channel[] getAltoChannels();
```
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/connection/ClientConnection.java:63]: 🤔 Are we sure this change is from this PR? Do we need a rebase?

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/ExperimentalAuthenticationMessageTask.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * Default Authentication with username password handling task
+ */
+public class ExperimentalAuthenticationMessageTask
```
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/ExperimentalAuthenticationMessageTask.java:34]: For this and custom message task: 

maybe we can reuse some code with non-experimental tasks.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AuthenticationMessageTask.java [hunk 4]
```diff
@@ -40,10 +40,8 @@ public AuthenticationMessageTask(ClientMessage clientMessage, Node node, Connect
     @Override
```
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AuthenticationMessageTask.java:40]: I did not get the changes in this file and customAuthMessageTask in general.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AuthenticationBaseMessageTask.java [hunk 5]
```diff
@@ -168,31 +168,29 @@ private AuthenticationStatus verifyEmptyCredentialsAndClusterName(PasswordCreden
...
-        return encodeAuth(status, null, null, (byte) -1, -1, null, clientFailoverSupported, false, null);
+        return encodeAuth(status, null, null, (byte) -1, "", -1, null, failoverSupported, null);
     }
 
     private ClientMessage prepareNotAllowedInCluster() {
-        boolean clientFailoverSupported = nodeEngine.getNode().getNodeExtension().isClientFailoverSupported();
+        boolean failoverSupported = nodeEngine.getNode().getNodeExtension().isClientFailoverSupported();
         byte status = NOT_ALLOWED_IN_CLUSTER.getId();
-        return encodeAuth(status, null, null, (byte) -1, -1, null, clientFailoverSupported, false, null);
+        return encodeAuth(status, null, null, (byte) -1, "", -1, null, failoverSupported, null);
```
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AuthenticationBaseMessageTask.java:182]: should not we remove altoPorts field instead 🤔 ?

**Review by srknzl — COMMENTED** [review]
Looks good other than the comments I left.

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/ExperimentalAuthenticationMessageTask.java (hunk 3 above)
Comment (srknzl) [reply hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/ExperimentalAuthenticationMessageTask.java:34]: maybe the following part can be moved to the baseMessageTask:

```java
assert parameters.uuid != null;
        clientUuid = parameters.uuid;
        clusterName = parameters.clusterName;
        credentials = new UsernamePasswordCredentials(parameters.username, parameters.password);
        clientSerializationVersion = parameters.serializationVersion;
        clientVersion = parameters.clientHazelcastVersion;
        clientName = parameters.clientName;
        labels = Collections.unmodifiableSet(new HashSet<>(parameters.labels));
```

**Review by OzanCansel — APPROVED** [review]

**Comment by mdumandag** [comment]
run-lab-run

**Pull request merged**